### DISCUSSION TITLE
More general catch

### DISCUSSION THREAD

**Discussion by OdinAnDieFreude** [dotnet/csharplang#5345] [Language Ideas]
It would be nice to be able to have a catch after every possible statement or code block.

Instead of 

    try
    {
      //some lines of code
    }
    catch (Exception ex)
    {
      //handle exception
    }

You could also do

    {
      //some lines of code
    }
    catch (Exception ex)
    {
      //handle exception
    }

That doesn't save a lot, but the more interesting part would be with methods.

Instead of

    public void MyMethod()
    {
      try
      {
        //some lines of code
      }
      catch (Exception ex)
      {
        //handle exception 
      }
    }

you could have

    public void MyMethod()
    {
      //some lines of code
    }
    catch (Exception ex)
    {
      //handle exception
    }

Of course, the catch would not have access to local variables of the method, but it would have access to the parameters provided:

    public void MyMethod(string test)
    {
      //some lines of code
    }
    catch (Exception ex)
    {
      handleException(test);
    }

Also, the same would be possible with other constructs:

before:

    if (<condition>)
    {
      try
      {
        //code
      }
      catch (Exception ex)
      {
        //handle exception
      }
    }

after:

    if (<condition>)
    {
      //code
    }
    catch (Exception ex)
    {
      //handle exception
    }

before:

    while (<condition>)
    {
      try
      {
        //code
      }
      catch (Exception ex)
      {
        //handle exception
      }
    }

after:

    while (<condition>)
    {
      //code
    }
    catch (Exception ex)
    {
      //handle exception
    }

Also try-catch without curly braces would be nice, both for the try and also for the catch.

    try SomeMethod();
    catch (Exception ex)
      handleException(ex);

or

    SomeMethod();
    catch (Exception ex)
      handleException(ex);

For variable declarations in a statement, the variable would be in the outer scope:

the following:

    try int val = SomeMethod();
    catch (Exception ex)
      handleException(ex);

would be equal to

    int val;
    try
    {
      val = SomeMethod();
    }
    catch (Exception ex)
    {
      handleException();
    }

But it would also work with the var keyword:

    try var val = SomeMethod();
    catch (Exception ex)
      handleException(ex);
    }

Of course, this is not true with a code block:

    {
      int val = SomeMethod();
    }
    catch (Exception ex)
      handleException(ex);

would be the same as

    try
    {
      int val = SomeMethod();
    }
    catch (Exception ex)
    {
      handleException(ex);
    }

Some more examples:

    if (condition)
      doSomething();
      catch (Exception ex)
        handleError(ex);

would be equal to

    if (condition)
    {
      try
      {
        doSomething();
      }
      catch (Exception ex)
      {
        handleError(ex);
      }
    }

And

    switch (someMethod())
    {
      case 1:
        doSomething1();
        break;
      case 5:
        doSomething5();
        break;
    }
    catch (Exception ex)
      handleException(ex);

would be equal to

    var x = someMethod();
    try
    {
      switch (x)
      {
        case 1:
          doSomething1();
          break;
        case 5:
          doSomething5();
          break;
      }
    }
    catch (Exception ex)
    {
      handleException(ex);
    }

and with lambdas:

    Func<int, int, int> f = (x, y) => x*x + y;
    catch (Exception ex)
      handleException(ex);

would be the same as

    Func<int, int, int> f;
    try
    {
      f = (x, y) => x*x + y;
    }
    catch (Exception ex)
      handleException(ex);

so nothing special, just an assignment.

But

    Func<int, int, int> f = (x, y) =>
    {
      return x*x + y;
      catch (Exception ex)
        handleException(ex);
    };

would catch inside the lambda.

With of course means the catch also works for the return statement.

The try keyword would now basically be without neccessity, but it would enforce a catch

The following would be legal

    DoSomething();
    catch (Exception ex)
      handleException(ex);

But the following wouldn't

    try DoSomething();
    try DoSomethingElse();
    catch (Exception ex)
      handleException(ex);

because the first try has no corresponding catch.

**Comment by gulshan** [comment]
Similar to #4543

**Reply by OdinAnDieFreude** [reply]
Not really.

Take the following code:

    void MyFunc()
    {
      OperationA();
      OperationB();
      catch (MyException1 me1)
      {
        Log(me1)
      }
      catch (MyException2 me2)
      {
        Log(me2);
      }
    }

In #4543, it would be equivalent to

    void MyFunc()
    {
      try
      {
        OperationA();
        OperationB();
      }
      catch (MyException1 me1)
      {
        Log(me1)
      }
      catch (MyException2 me2)
      {
        Log(me2);
      }
    }

But in my proposal, it would be equivalent of

    void MyFunc()
    {
      OperationA();
      try
      {
        OperationB();
      }
      catch (MyException1 me1)
      {
        Log(me1)
      }
      catch (MyException2 me2)
      {
        Log(me2);
      }
    }

And the code intended by #4543's example would be

    void MyFunc()
    {
      OperationA();
      OperationB();
    }
    catch (MyException1 me1)
    {
      Log(me1);
    }
    catch (MyException2 me2)
    {
      Log(me2);
    }

or even

    void MyFunc()
    {
      OperationA();
      OperationB();
    }
    catch (MyException1 me1) Log(me1);
    catch (MyException2 me2) Log(me2);


Also, I propose the same consistently for every block or statement.
